lib/mix/tasks/burn.ex

```diff
@@ -71,56 +76,39 @@ defmodule Mix.Tasks.Burn do
         dev -> dev
       end
 
+    task = if opts[:overwrite], do: "complete", else: opts[:task] || "upgrade"
+
     set_provisioning(firmware_config[:provisioning])
-    burn(fw, dev, opts, argv)
+
+    burn(fw, dev, task, argv)
 
     # Remove the temporary .fw file
     WSL.cleanup_file(fw, firmware_location)
   end
 
-  defp burn(fw, dev, opts, argv) do
-    task = opts[:task] || "complete"
+  defp burn(fw, dev, task, argv) do
     args = ["-a", "-i", fw, "-t", task, "-d", dev] ++ argv
 
-    {cmd, args} =
-      case :os.type() do
-        {_, :darwin} ->
-          {"fwup", args}
-
-        {_, :linux} ->
-          if WSL.running_on_wsl?() do
-            WSL.admin_powershell_command("fwup", Enum.join(args, " "))
-          else
-            fwup = System.find_executable("fwup")
-
-            case File.stat(dev) do
-              {:ok, %File.Stat{access: :read_write}} ->
-                {"fwup", args}
-
-              {:error, :enoent} ->
-                case File.touch(dev, System.os_time(:second)) do
-                  :ok ->
-                    {"fwup", args}
-
-                  {:error, :eacces} ->
-                    elevate_user()
-                    {"sudo", provision_env() ++ [fwup] ++ args}
-                end
-
-              _ ->
-                elevate_user()
-                {"sudo", provision_env() ++ [fwup] ++ args}
-            end
-          end
-
-        {_, :nt} ->
-          {"fwup", args}
-
-        {_, type} ->
-          raise "Unable to burn firmware on your host #{inspect(type)}"
-      end
+    os = get_os!()
 
-    shell(cmd, args)
+    {cmd, args} = cmd_and_args_for_os(os, args, dev)
+
+    shell(cmd, args, stream: FwupStream.new())
+    |> format_result(task)
+    |> case do
+      :failed_not_upgradable ->
+        Mix.shell().info("""
+        #{IO.ANSI.yellow()}
+        FWUP upgrade was attempted but #{dev} does not have a previous firmware.
+        Switching to use the #{IO.ANSI.cyan()}complete#{IO.ANSI.yellow()} task to format the entire card.
```

Comment:
```suggestion
        Device #{dev} either doesn't have firmware on it or has incompatible firmware.
        Going to burn the whole MicroSD card so that it's in a factory-default state.
```